src/OrchardCore/OrchardCore.FileStorage.AzureBlob/BlobFileStore.cs

```diff
@@ -0,0 +1,286 @@
+            // listing directories.
+            var placeholderBlob = GetBlobReference(this.Combine(path, _directoryMarkerFileName));
+            await placeholderBlob.UploadTextAsync("This is a directory marker file created by Orchard Core. It is safe to delete it.");
+
+            return true;
+        }
+
+        public async Task<bool> TryDeleteFileAsync(string path)
+        {
+            await _verifyContainerTask;
+
+            var blob = GetBlobReference(path);
+
+            return await blob.DeleteIfExistsAsync();
+        }
+
+        public async Task<bool> TryDeleteDirectoryAsync(string path)
+        {
+            if (String.IsNullOrEmpty(path))
+                throw new FileStoreException("Cannot delete the root directory.");
+
+            await _verifyContainerTask;
+
+            var blobDirectory = GetBlobDirectoryReference(path);
+
+            BlobContinuationToken continuationToken = null;
+            var blobsWereDeleted = false;
+
+            do
+            {
+                var segment = 
+                    await blobDirectory.ListBlobsSegmentedAsync(
+                        useFlatBlobListing: true, 
+                        blobListingDetails: BlobListingDetails.None, 
+                        maxResults: null, 
+                        currentToken: continuationToken, 
+                        options: null, 
+                        operationContext: null);
+
+                foreach (var item in segment.Results)
+                {
+                    if (item is CloudBlob blob)
+                    {
+                        await blob.DeleteAsync();
+                        blobsWereDeleted = true;
+                    }
+                }
+
+                continuationToken = segment.ContinuationToken;
+            }
+            while (continuationToken != null);
+
+            return blobsWereDeleted;
+        }
+
+        public async Task MoveFileAsync(string oldPath, string newPath)
+        {
+            await _verifyContainerTask;
+
+            await CopyFileAsync(oldPath, newPath);
+            await TryDeleteFileAsync(oldPath);
+        }
+
+        public async Task CopyFileAsync(string srcPath, string dstPath)
+        {
+            if (srcPath == dstPath)
+                throw new ArgumentException($"The values for {nameof(srcPath)} and {nameof(dstPath)} must not be the same.");
+
+            await _verifyContainerTask;
+
+            var oldBlob = GetBlobReference(srcPath);
+            var newBlob = GetBlobReference(dstPath);
+
+            if (!await oldBlob.ExistsAsync())
+                throw new FileStoreException($"Cannot copy file '{srcPath}' because it does not exist.");
+
+            if (await newBlob.ExistsAsync())
+                throw new FileStoreException($"Cannot copy file '{srcPath}' because a file already exists in the new path '{dstPath}'.");
+
+            var operationId = await newBlob.StartCopyAsync(oldBlob);
```

Comment:
Is that a common pattern? there is no `newBlob.CopyAsync`, and no timeout?

Reply:
I too find it very weird. Very strange that you can't await the copy operation. But this is the only way I could find in the SDK.